### DISCUSSION TITLE
Understanding nested route rendering

### DISCUSSION THREAD

**Discussion by vaneenige** [remix-run/remix#2178] [Q&A] [answered]
[Nested routes](https://remix.run/docs/en/v1/guides/routing#what-are-nested-routes) is looking like the solution we've been looking for, but there are some questions:

We have the following structure (which allows us to have separate loaders for language related things and pages):
- `$lang`
  - `about.tsx`
  - `index.tsx`
- `$lang.tsx`

It works as expected on first page load. When we go to `/en/about` everything is rendered once.
When we navigate we from `/en` to `/en/about` we thought only that `<Outlet />` would render again, but it looks like the whole page does (and multiple times due to the transition state changing).

Here's what we logged:

On page load: ✅ 
```
Render: `/en` ($lang)
Render: `/en` (index)
```

On navigate: 🔴 
```
Render: `/en` ($lang)
Render: `/en` (index)
Render: `/en` ($lang)
Render: `/en` (index)
Render: `/en` ($lang)
Render: `/en/about`
```

Here are my questions: 
- Is there a way to only have the `<Outlet />` render (in this case everything inside of the `$lang` folder and not `$lang.tsx`)?
- Is there a way to turn off the multiple renders caused by the transition when we don't need it?

**Comment by sergiodxa** [comment, marked as answer]
You can probably use React.memo, but a lot will depend on what hooks you are calling inside your routes, if you are calling useTransition then your component will re render on any transition change.

I would recommend you to don’t care that much about re render, if they are slow it’s better to make them fast than avoiding re renders.

**Reply by vaneenige** [reply]
Thank you for the answer! Wouldn't avoiding unnecessary renders always be a good thing?
Even when you don't use `useTransition` the extra renders are done for it which is odd I think. 

Not going to mess around with `React.memo` as performance is good, just wondering about structural improvements!

**Reply by sergiodxa** [reply]
Re-render is just how React works, and it doesn't mean it's going to update the DOM again, if your renders (calling the component body) is fast (and it should) then doing multiple of re-renders will not be an issue. The slow part is the commit to the DOM.

@kentcdodds has a great article on this https://kentcdodds.com/blog/fix-the-slow-render-before-you-fix-the-re-render.

**Reply by vaneenige** [reply]
I've seen that article a few times, a must read for every React developer! 💯  

Like I mentioned, performance is good and rendering is fine (as the DOM doesn't change), but at it's core I think it's good to limit work that can be avoided. It's also fine if it can't be disabled, technical choices are always interesting.
